src/jit/lowerxarch.cpp

```diff
@@ -1030,6 +1034,23 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
     GenTreePtr shiftBy = tree->gtOp.gtOp2;
     GenTreePtr source  = tree->gtOp.gtOp1;
 
+#ifdef _TARGET_X86_
+    // The first operand of a GT_LSH_HI and GT_RSH_LO oper is a GT_LONG so that
+    // we can have a three operand form. Increment the srcCount.
+    if (source->OperGet() == GT_LONG)
+    {
+        info->srcCount++;
+        source->gtLsraInfo.isDelayFree = true;
+        GenTreePtr sourceLo = source->gtOp.gtOp1;
+        GenTreePtr sourceHi = source->gtOp.gtOp2;
+        sourceLo->gtLsraInfo.isDelayFree = true;
+        sourceHi->gtLsraInfo.isDelayFree = true;
+
+        source->gtLsraInfo.hasDelayFreeSrc = true;
+        info->hasDelayFreeSrc = true;
+    }
```

Comment:
HasDelayFreeSrc setting looks correct to me.
Only which of the lo or hi sources gets marked as isDelayFree could be optimized.